### ISSUE TITLE
make parameter passing to shell scripts secure

### ISSUE THREAD

**Issue by JensLincke** [LivelyKernel/lively4-server#7]
Is there an accepted general way to pass parameters to the scripts securely? I just stripping our \' enough?

```JS
  var repository = req.headers["gitrepository"]
      var msg = req.headers["gitcommitmessage"].replace(/'/g,"")
      // # TODO...
      var cmd = 'cd ' + repository + "; git commit -m '"+ msg +"' -a "
```

**Comment by JensLincke** [comment]
@timfel @fniephaus

**Comment by fniephaus** [comment]
In this example, I'd make sure that repository can only be a valid path/directory (allowed chars are `[a-z]/.~`) and msg can only be a normal text without special characters. In the above code, I could set `gitrepository` to e.g. `~/; sudo whoami`, I'm sure I could find something similar for `gitcommitmessage`.

**Comment by jgraichen** [comment]
Good programming languages usually have something like `exec(exe, *args)`. A fork/exec call that is not spawning a shell but directly invoking the executable (git) and passing `args` as the usual `argv` array in `main(char* argv)`.

**Comment by jgraichen** [comment]
https://nodejs.org/api/child_process.html

**Comment by jgraichen** [comment]
No. Something like `exec('git', 'commit', '-m', params["message"])` would always pass the param as the message string to git. That is only exploitable if git itself has a sanitizing bug.

**Comment by fniephaus** [comment]
Passing the responsibility to the target command is probably a good idea here. And yes, it looks like `exec` has protection built in.

**Comment by JensLincke** [comment]
but.. since the target programming is a shell script that we wrote... e.g. lively4sync than we have to deal with the security ourselves there again.... some of the commands are not just direct calls to the git api.

**Comment by fniephaus** [comment]
Rewrite the shell script in JS? ;)

**Comment by JensLincke** [comment]
if it would have been easy to write in JS I would not have needed a shell script ;-)

**Comment by JensLincke** [comment]
but for some I see that this is the most secure option...

**Comment by fniephaus** [comment]
let have a loot at this next monday!

**Comment by JensLincke** [comment]
Tim sagt du sollst JETZT  Wochenende machen.... 
Nicole ist doch da!

**Comment by fniephaus** [comment]
Tim ist d##f :wink: selber!
